Realm/RLMManagedDictionary.mm

```diff
@@ -0,0 +1,546 @@
+    // TODO: implement hash
+    //return std::hash<realm::object_store::Dictionary>()(_backingCollection);
+    return 0;
+}
+
+- (NSUInteger)countByEnumeratingWithState:(NSFastEnumerationState *)state
+                                  objects:(__unused __unsafe_unretained id [])buffer
+                                    count:(NSUInteger)len {
+    return RLMFastEnumerate(state, len, self);
+}
+
+#pragma mark - Object Retrieval
+
+- (nullable id)objectForKey:(id<RLMDictionaryKey>)key {
+    try {
+        RLMAccessorContext context(*_objectInfo);
+        return _backingCollection.get(context,
+                                      keyFromRLMDictionaryKey(key, context));
+    }
+    catch (realm::KeyNotFound const&) {
+        return nil;
+    }
+    catch (...) {
+        throwError<RLMManagedDictionary>(nil, nil);
+    }
+}
+
+- (nullable id)objectForKeyedSubscript:(id<RLMDictionaryKey>)key {
+    return [self objectForKey:key];
+}
+
+- (nonnull id)objectAtIndex:(NSUInteger)index {
+    return translateErrors<RLMManagedDictionary>([&] {
+        auto key = _backingCollection.get_pair(index).first;
+        return RLMStringDataToNSString(key);
+    });
+}
+
+- (NSUInteger)indexOfObject:(id)value {
+    return translateErrors<RLMManagedDictionary>([&] {
+        return _backingCollection.find_any(value);
+    });
+}
+
+- (void)setObject:(id)obj forKey:(id<RLMDictionaryKey>)key {
+    RLMDictionaryValidateMatchingObjectType(self, key, obj);
+    changeDictionary(self, ^{
+        RLMAccessorContext context(*_objectInfo);
+        _backingCollection.insert(context,
+                                  keyFromRLMDictionaryKey(key, context),
+                                  obj);
+    });
+}
+
+- (void)setObject:(id)obj forKeyedSubscript:(id<RLMDictionaryKey>)key {
+    // passing `nil` to the subscript should delete the object.
+    if (!obj) {
+        [self removeObjectForKey:key];
+        return;
+    }
+    RLMDictionaryValidateMatchingObjectType(self, key, obj);
+    changeDictionary(self, ^{
+        RLMAccessorContext context(*_objectInfo);
+        _backingCollection.insert(context,
+                                  keyFromRLMDictionaryKey(key, context),
+                                  obj);
+    });
+}
+
+- (void)removeAllObjects {
+    changeDictionary(self, ^{
+        _backingCollection.remove_all();
+    });
+}
+
+- (void)removeObjectsForKeys:(NSArray *)keyArray {
+    changeDictionary(self, ^{
+        RLMAccessorContext context(*_objectInfo);
+        for (id key in keyArray) {
+            _backingCollection.erase(keyFromRLMDictionaryKey(key, context));
+        }
+    });
+}
+
+- (void)removeObjectForKey:(id<RLMDictionaryKey>)key {
+    changeDictionary(self, ^{
+        RLMAccessorContext context(*_objectInfo);
+        _backingCollection.erase(keyFromRLMDictionaryKey(key, context));
+    });
+}
+
+inline realm::StringData keyFromRLMDictionaryKey(id<RLMDictionaryKey> key, RLMAccessorContext &context) {
```

Comment:
This function doesn't really make any sense for string-only keys and also doesn't fit with the future design for other key types.